### DISCUSSION TITLE
Blacklist doesn't exist in GPO error message

### DISCUSSION THREAD

**Discussion by Andrew-Griggs** [Romanitho/Winget-AutoUpdate#640] [Q&A]
Hi all, I'm testing various configurations and I can't seem to get the blacklist working.  Every update cycle throws the following error:

################################################################
#     21/06/2024 - CHECK FOR APP UPDATES (System context - Connected user)
################################################################
18:52:53 - WAU Policies management Enabled.
18:52:53 - Notification Level: Full. Notification Language: English
18:52:53 - Checking internet connection...
18:52:53 - Connected !
18:52:53 - Winget Version: v1.7.11261
18:52:53 - WAU current version: 1.20.1
18:52:53 - WAU AutoUpdate is Enabled.
18:52:54 - WAU is up to date.
18:52:54 - WAU uses External Lists from: GPO
18:52:54 - WAU uses Black List config
18:52:54 - Critical: Blacklist doesn't exist in GPO, exiting...

I can't see what I'm missing in my config though, here's the policies downloaded from Intune:

![image](https://github.com/Romanitho/Winget-AutoUpdate/assets/75014604/ae365169-cee0-465a-ba45-850efb2345e5)

![image](https://github.com/Romanitho/Winget-AutoUpdate/assets/75014604/d0b1064e-5665-4603-b823-e3e5cbdb074f)

What am I missing?  Interestingly, if I switch to a whitelist approach everything works perfectly.

**Comment by Romanitho** [comment]
@KnifMelti  any idea ?
I have used blacklist without any pb. I'll check when I have few minutes

**Reply by KnifMelti** [reply]
Latest change maybe?:
https://github.com/Romanitho/Winget-AutoUpdate/pull/580

**Reply by Andrew-Griggs** [reply]
Just to add, if I disable the GPO management and switch to Configurator, everything seems to work as normal, but then switching back to GPO breaks it again.

**Reply by KnifMelti** [reply]
> Just to add, if I disable the GPO management and switch to Configurator, everything seems to work as normal, but then switching back to GPO breaks it again.

What is _**Configurator**_?
Picture?

**Reply by KnifMelti** [reply]
Here's my run with Black List in GPO (after via `_WAU-mods.ps1` enabling and running the task **Winget-AutoUpdate-Policies** everytime WAU runs. As standard it only runs once a day at 06:00 evaluating the settings in GPO) with the latest WAU.
Nothing wrong with it.

Can it be you're using the **fork** https://github.com/Weatherlights/Winget-AutoUpdate-Intune?
In that case you should ask about this in that repo.

![image](https://github.com/Romanitho/Winget-AutoUpdate/assets/102996177/7c34ba05-b8ef-48d6-8f1b-576655f11703)

![image](https://github.com/Romanitho/Winget-AutoUpdate/assets/102996177/48709b5b-1efb-44c7-852a-84ab9968ebf9)

**Reply by Andrew-Griggs** [reply]
I'm referring to this Configurator:
![image](https://github.com/Romanitho/Winget-AutoUpdate/assets/75014604/00dd16d2-1c3d-49b4-af56-1c26a051d431)
I'm not using that fork, I've got all files direct from here: https://github.com/Romanitho/Winget-AutoUpdate

**Reply by KnifMelti** [reply]
Then, when you have the GPO in registry, run  the task Winget-AutoUpdate-Policies (it only runs once a day at 06:00).
That should create an registry entry in the WAU uninstall section an a text file under logs
When WAU is run after that things should work.
@Romanitho is the creator of this handling.

**Reply by KnifMelti** [reply]
IE  (at my computer now):
- `WAU-Policies.ps1` is run by the task `Winget-AutoUpdate-Policies` (it only runs once a day at 06:00 or later if that time is missed)
- `WAU-Policies.ps1` checks if `HKLM:\SOFTWARE\Policies\Romanitho\Winget-AutoUpdate\WAU_ActivateGPOManagement = 1`
- If true then `HKLM:\SOFTWARE\Microsoft\Windows\CurrentVersion\Uninstall\Winget-AutoUpdate\WAU_RunGPOManagement` is set to `1`
- The result is saved to `logs\LatestAppliedSettings.txt`
- `Winget-Upgrade.ps1` takes care of the list in GPO

**Reply by KnifMelti** [reply]
Your policies registry should look something like this (and your BlackList folder like you have it):

![image](https://github.com/Romanitho/Winget-AutoUpdate/assets/102996177/bac45d6b-69ce-41cf-8a72-4acbb284d2fa)

Your settings registry should look something like this:

![image](https://github.com/Romanitho/Winget-AutoUpdate/assets/102996177/75144625-b74e-4276-8d1b-247b2775caae)

**Comment by KnifMelti** [comment]
Found the culprit!
It's because you only have 1 entry in the blacklist...
...add another and it's fine - and report a bug, please!

**Reply by KnifMelti** [reply]
The solution is to in `Winget-Upgrade.ps1` avoid the `$toUpdate/$toSkip.GetUpperBound(0)` and use this snippet instead (**[Create issue from discussion](https://github.com/Romanitho/Winget-AutoUpdate/issues/new?created_from_discussion_number=640)**):
```
if ($UseWhiteList) {
    if (-not $toUpdate) {
        Write-ToLog "Critical: Whitelist doesn't exist in GPO, exiting..." "Red"
        New-Item "$WorkingDir\logs\error.txt" -Value "Whitelist doesn't exist in GPO" -Force
        Exit 1
    }
    foreach ($app in $toUpdate) { Write-ToLog "Include app ${app}" }
}
else {
    if (-not $toSkip) {
        Write-ToLog "Critical: Blacklist doesn't exist in GPO, exiting..." "Red"
        New-Item "$WorkingDir\logs\error.txt" -Value "Blacklist doesn't exist in GPO" -Force
        Exit 1
    }
    foreach ($app in $toSkip) { Write-ToLog "Exclude app ${app}" }
}

```

**Reply by KnifMelti** [reply]
PR #648
